dialects/postgresql/src/main/kotlin/app/cash/sqldelight/dialects/postgresql/grammar/PostgreSql.bnf

```diff
@@ -102,14 +102,15 @@ type_name ::= (
   override = true
 }
 bind_parameter ::= ( DEFAULT | '?' | ':' {identifier} ) {
```

Comment:
should we remove `'?'` here then?

Reply:
I would keep it! JDBC supports it, by replacing it under the hood, and for syntactical sugar, it is nice.
But custom dialects could drop this support: eg my postgres native driver :D https://github.com/hfhbd/postgres-native-sqldelight/issues/50

Reply:
This also has implications for the included R2DBC driver, since R2DBC doesn't replace the `?` params like JDBC does. It only accepts the `$1` placeholders.

That said, I don't know of anyone who has any intentions of using that driver or if we even want to keep the driver in SQLDelight itself.

Reply:
Thanks, good to know. I think, it should be possible to overwrite the bind parameter for this driver/dialect then :D

Reply:
@dellisd Would adding `'$'` be enough? I don't want to create a new dialect just for R2DBC...

Reply:
Nice, #3168 added the 'integration-mysql-async' test project but no test to execute it :D

Reply:
So this query does not work with R2DBC: `INSERT INTO dog VALUES (?, ?, ?)` and needs to be rewritten as `INSERT INTO dog VALUES ($1, $2, $3)`?

Reply:
The R2DBC MySQL driver is currently (still) broken because of a binary incompatibility so it's not possible to fully test it.

The thing about R2DBC is that it doesn't enforce any one placeholder character like JDBC does. So `?` works fine with the MySQL driver, but the postgres and h2 drivers expect the `$1` form. As far as I can tell, they require the position too. So `($, $, $)` wouldn't work. It needs to be `($1, $2, $3)`. I haven't actually tested this though.

I wouldn't worry about it too much though. If it turns out someone _actually needs_ the R2DBC drivers, we can revisit this at that time.

Reply:
What do you mean by broken? Yeah, it took some time to create a working sample, I will upload it sone with a new issue about the IMHO confusing async runtime, because there is no one :)
My current implementation allows rewriting the actual queries for different dialects and drivers (aka generateAsync).

Reply:
The last release of the MySQL R2DBC driver was built against an older version of `r2dbc-spi`, and if I recall correctly the return type of the execute method (or whatever it was called) was changed from `Integer` to `Long`, so trying to execute any query using the new `spi` and the MySQL driver results in a `ClassCastException`.

Being able to rewrite it in a dialect sounds fine to me. Like I said, we can deal with R2DBC use cases **if** they ever pop up. :wink:

Reply:
Hm, I created https://github.com/cashapp/sqldelight/issues/3523 to track the async implementation but I didn't get a CCE. Maybe because https://github.com/mirromutth/r2dbc-mysql does not use the latest spi version too :D

Reply:
Yeah, you are right about the CCE, so I downgraded r2dbc to `0.9.1-RELEASE`.

Reply:
I would normally agree, but at the moment, the two usage of this driver are mysql and postgresql. Both implementation don't support 1.0.0-RELEASE, so upgrading to this version breaks the usage. I will update/revert the downgrade after a new release of one of the drivers.

Reply:
I think I agree with that justification, I'd rather have a working version of the driver checked in, and then we can start tracking updates as frequently as we can